staging/src/k8s.io/apiserver/pkg/endpoints/handlers/responsewriters/writers.go

```diff
@@ -245,11 +274,23 @@ func (w *deferredResponseWriter) Write(p []byte) (n int, err error) {
 	return w.w.Write(p)
 }
 
-func (w *deferredResponseWriter) Close() error {
+func (w *deferredResponseWriter) Close() (err error) {
 	if !w.hasWritten {
-		return nil
+		if !w.hasBuffered {
+			return nil
+		}
+		_, writeErr := w.unbufferedWrite(w.buffer)
+		w.buffer = nil
+		if writeErr != nil {
+			defer func() {
+				if err == nil {
+					// if there was an error writing the buffer, return that rather than a nil return value
+					err = writeErr
+				}
+			}()
+		}
```

Comment:
```suggestion
	_, err := w.unbufferedWrite(w.buffer)
	w.buffer = nil
	return err
```

If we haven't written anything, all the data should be in buffer and we didn't use use gzip, so we can just return.

Reply:
I agree, but the code lines 279 to 281 should not be removed. If there is neither `writing` nor `buffering`, it should return directly to ensure consistency with the original logic.

```go
func (w *deferredResponseWriter) Close() (err error) {
	if !w.hasWritten {
		// should not be removed
		if !w.hasBuffered {
			return nil
		}
        //...
}
```

Reply:
Right!

Reply:
Could you maybe add a test for that?